### PR TITLE
Add deleteEmptyFolder attribute and create action to fileSystemActor

### PR DESCRIPTION
**Pull request by alisihab**

### CODE REVIEW

File: core/src/main/java/nl/nn/adapterframework/filesystem/IBasicFileSystem.java [hunk 1]
```diff
@@ -47,21 +47,21 @@
...
 	public boolean isOpen();
 
-
 	/**
-	 * Lists all files in 'folder' or in the 'root' of the filesystem (when folder is null). 
+	 * Lists all files in 'folder' or in the 'root' of the filesystem (when folder is null).
 	 * Should list only 'files', no folders.
 	 */
-	public DirectoryStream<F> listFiles(String folder) throws FileSystemException;
+	public DirectoryStream<F> listFiles(String folder, boolean ignoreFolders) throws FileSystemException;
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/filesystem/IBasicFileSystem.java:54]: There should not be a new parameter `ignoreFolders`, listFiles should never return folders.

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/test/java/nl/nn/adapterframework/filesystem/BasicFileSystemTest.java [hunk 2]
```diff
@@ -498,4 +498,42 @@ public void basicFileSystemTestListFileShouldNotReadFolders() throws Exception {
...
+
+	@Test
+	public void getParentOfTheDeletedFolder() throws Exception {
+		String folderName = "parentFolder";
+
+		fileSystem.configure();
+		fileSystem.open();
+
+		_createFolder(folderName);
+		_createFolder(folderName+"/innerFolder");
```
Comment (gvanbrakel) [inline core/src/test/java/nl/nn/adapterframework/filesystem/BasicFileSystemTest.java:528]: This just creates a folder *named* '<folderName>/innerFolder', not a sub folder of <folderName>

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/filesystem/FileSystemActor.java [hunk 3]
```diff
@@ -531,6 +547,27 @@ private void writeContentsToFile(OutputStream out, Message input, ParameterValue
...
+	private void deleteEmptyFolder(F f) throws FileSystemException, IOException {
+		if(isDeleteEmptyFolder()) {
+			deleteEmptyFolder(fileSystem.getParentFolder(f));
+		}
+	}
+
+	private void deleteEmptyFolder(String folder) throws FileSystemException, IOException {
+		if(isDeleteEmptyFolder()) {
+			boolean isEmpty = false;
+			try (Stream<F> stream = FileSystemUtils.getFilteredStream(fileSystem, folder, null, null)) {
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/filesystem/FileSystemActor.java:559]: When there are no wildcards, you can just use fileSystem.listFiles(folder)

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/filesystem/Samba2FileSystem.java [hunk 4]
```diff
@@ -163,65 +164,64 @@ public void close() throws FileSystemException {
 	private AuthenticationContext authenticate() throws FileSystemException {
 		CredentialFactory credentialFactory = new CredentialFactory(getAuthAlias(), getUsername(), getPassword());
 		if (StringUtils.isNotEmpty(credentialFactory.getUsername())) {
-			if(StringUtils.equalsIgnoreCase(authType, "NTLM")) {
+			if(authType == Samba2AuthType.NTLM) {
 				return new AuthenticationContext(getUsername(), password.toCharArray(), getDomain());
-			}else if(StringUtils.equalsIgnoreCase(authType, "SPNEGO")) {
+			}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/filesystem/Samba2FileSystem.java:169]: Please use `switch()` for choosing an action based on a enum

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/filesystem/Samba2FileSystem.java [hunk 5]
```diff
@@ -163,65 +164,67 @@ public void close() throws FileSystemException {
...
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/filesystem/Samba2FileSystem.java:169]: Now the switch is there, the 'if' can be removed

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/filesystem/FtpFileSystem.java [hunk 6]
```diff
@@ -281,6 +281,12 @@ public String getName(FTPFile file) {
 		}
 		return file.getName();
 	}
+	
+	@Override
+	public String getParentFolder(FTPFile file) throws FileSystemException {
+		FTPFileRef fileRef = (FTPFileRef) file;
+		return fileRef.getFolder();
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/filesystem/FtpFileSystem.java:288]: I don't think you can just cast this.

Comment (alisihab) [reply core/src/main/java/nl/nn/adapterframework/filesystem/FtpFileSystem.java:288]: before this method is called with file toFile method is called and that returns FtpFileRef but I agree this is confusing

**Review by nielsm5 — APPROVED** [review]

**Pull request merged**